### PR TITLE
54 mesido validation with pandapipes simple model

### PR DESCRIPTION
**Pull request by KobusVanRooyen**
Convert the pandapipes validation stuff into a test case

Still to complete:

- [x] formatting, exclude pandapipes parser file from formatting checks
- [x] getting pipeline working with pandapipes (install goodies needed etc)

### CODE REVIEW

File: tests/test_validation_pandapipes.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+        # ------------------------------------------------------------------------------
+        # Setup and run pandapipes
+        # ------------------------------------------------------------------------------
+        # ------------------------------------------------------------------------------
+        from examples.pandapipes.src.pandapipeesdlparser import PandapipeEsdlParserClass
+
+        esdl_file = os.path.join(base_folder, "model", "Test_Tree_S1C1.esdl")
+        esdlparser = PandapipeEsdlParserClass()
+        esdlparser.loadESDLFile(esdl_file)
+        esdlparser.add_additional_joint()
```
Comment (jimrojerTNO) [inline tests/test_validation_pandapipes.py:99]: maybe a comment on why this line of code is needed

File: tests/test_validation_pandapipes.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+            net,
+            element="circ_pump_mass",
+            variable="t_flow_k",
+            element_index=net.circ_pump_mass.index.values[isupply],
+            data_source=df_supply_pump_temperature_kelvin,
+            profile_name=net.circ_pump_mass.index.values[isupply].astype(str),
+        )
+        supply_flow_kgs_s = pd.DataFrame(supply_flow, columns=[f"{isupply}"])
+        ds_supply_control_flow_kgs_s = DFData(supply_flow_kgs_s)
+        control.ConstControl(
```
Comment (jimrojerTNO) [inline tests/test_validation_pandapipes.py:160]: for me it is unclear why this additional ConstControl is needed? Is this not already set by the first ConstControl

File: tests/test_validation_pandapipes.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+        # Setup supply mass flow for panda_pipes
+        average_temperature_kelvin = (supply_temperature + return_temperature) / 2.0 + 273.15
+        cp_joule_kgkelvin = pp.get_fluid(net).get_heat_capacity(average_temperature_kelvin)
+
+        # Enforce mass flow rate instead of cacluting it from Q = m_dot...
+        mesido_demand_flow_kg_s = results["Pipe1.Q"] * 988.0
+        demand_flow = mesido_demand_flow_kg_s
+        supply_flow = demand_flow * total_consumers / total_producers
+
+        # Assign profiles to assets
```
Comment (jimrojerTNO) [inline tests/test_validation_pandapipes.py:129]: If I understand correctly the constControl is basically setting the boundary conditions. Can we in a comment state which boundary conditions we set and what states now follow from the simulation.

File: tests/test_validation_pandapipes.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+            (
+                ow.np_results["res_pipe.p_to_bar"][0:mdata_points]
+                - ow.np_results["res_pipe.p_from_bar"][0:mdata_points]
+            )
+            * 100.0e3
+            / density[0][0]
+            / GRAVITATIONAL_CONSTANT
+        )
+
+        # Compare head losses
```
Comment (jimrojerTNO) [inline tests/test_validation_pandapipes.py:245]: I would suggest that we include the plotting code (commented), so that people can locally run validation and visualize easily

File: tests/test_validation_pandapipes.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+import pandapower.control as control
+from pandapower.timeseries import DFData
+from pandapower.timeseries import OutputWriter
+
+import pandas as pd
+
+from utils_tests import demand_matching_test, energy_conservation_test, heat_to_discharge_test
+
+
+class ValidateWithPandaPipes(TestCase):
```
Comment (jimrojerTNO) [inline tests/test_validation_pandapipes.py:26]: This is an amazing test case.

To make it more complete I would suggest two minor efforts to make this more prominent.
- In the heat_physics.rst in docs add a paragraph that states that the headlosses are validated using pandapipes (ideally also include the graphs of the paper) and refer to this test case.
- In the README.md of our code base states: "with multiple linearization strategies to conservatively approximate the steady-state physics and financial models." I would suggest to add a sentence here: "See for example our head loss validation using pandapipes in test folder".

File: manual_run/testing_bugs/src/run_grow_edwin.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
+from pathlib import Path
```
Comment (jimrojerTNO) [inline manual_run/testing_bugs/src/run_grow_edwin.py:1]: I don't understand why this file is part of this Pull Request? Am I missing something?

File: examples/pandapipes/input/timeseries_constant _original.csv [hunk 7]
```diff
@@ -0,0 +1,46 @@
+DateTime,demand_1
```
Comment (jimrojerTNO) [inline examples/pandapipes/input/timeseries_constant _original.csv:1]: I think this file is not used by any of the runs, can we remove it?

File: examples/pandapipes/input/timeseries_constant.csv [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+2013-05-19 23:00:00,44444444.0
+2013-05-20 00:00:00,38888889.0
+2013-05-20 01:00:00,33333333.0
+2013-05-20 02:00:00,27777778.0
+2013-05-20 03:00:00,22222222.0
+2013-05-20 04:00:00,16666667.0
+2013-05-20 05:00:00,11111111.0
+2013-05-20 06:00:00,5555556.0
+2013-05-20 07:00:00,1.0
+2013-05-20 08:00:00,50000000.0
```
Comment (jimrojerTNO) [inline examples/pandapipes/input/timeseries_constant.csv:12]: Can we not simply delete everything from line 12 onwards?

**Review by jimrojerTNO — COMMENTED** [review]
I think this is a great PR, very high impact as we allow people to reproduce the validation that we did.

I have some minor comments about code that can maybe be removed and suggestions to make the test case more prominent by referring to it in the documentation. Code wise I think it all looks solid.

File: examples/pandapipes/input/timeseries_constant.csv (hunk 8 above)
Comment (KobusVanRooyen) [reply examples/pandapipes/input/timeseries_constant.csv:12]: Yes. Done

File: examples/pandapipes/input/timeseries_constant _original.csv (hunk 7 above)
Comment (KobusVanRooyen) [reply examples/pandapipes/input/timeseries_constant _original.csv:1]: Thanks. Deleted

File: tests/test_validation_pandapipes.py (hunk 1 above)
Comment (KobusVanRooyen) [reply tests/test_validation_pandapipes.py:99]: Agree. Comment has been added

File: tests/test_validation_pandapipes.py (hunk 3 above)
Comment (KobusVanRooyen) [reply tests/test_validation_pandapipes.py:129]: Comment has been added and imrpoved

File: tests/test_validation_pandapipes.py (hunk 2 above)
Comment (KobusVanRooyen) [reply tests/test_validation_pandapipes.py:160]: Deleted. More info if you are interested: The control.ConstControl with element="flow_control" is not needed for this simple case. It is however needed for when there is more than 1 producer/heat demand. So I deleted the code for this test. I used this in my other runs when I compared chess with pandapipes for ring networks with 2 heat demands and 2 producers. Then you need to specify these controls and also deactivate it for two assets (1 heat demand and 1 producer) to ensure the system is not over constraints. If I remember correctly.

File: tests/test_validation_pandapipes.py (hunk 5 above)
Comment (KobusVanRooyen) [reply tests/test_validation_pandapipes.py:26]: Nice. README has been updated

Comment (KobusVanRooyen) [reply tests/test_validation_pandapipes.py:26]: heat_physics.rst has also been updated

File: manual_run/testing_bugs/src/run_grow_edwin.py (hunk 6 above)
Comment (KobusVanRooyen) [reply manual_run/testing_bugs/src/run_grow_edwin.py:1]: This is funny. Mistake from my side. It has been removed

**Comment by KobusVanRooyen** [comment]
@jimrojerTNO  ready for 2nd review

File: tests/test_validation_pandapipes.py (hunk 2 above)
Comment (jimrojerTNO) [reply tests/test_validation_pandapipes.py]: check

File: tests/test_validation_pandapipes.py (hunk 4 above)
Comment (jimrojerTNO) [reply tests/test_validation_pandapipes.py:229]: thnx looks good

**Review by jimrojerTNO — APPROVED** [review]
Very good work!
